syntax = "proto3";

/*************************************交易相关协议*************************************/
/*************************************Transaction-related agreements*************************************/
import "sign.proto";


message SignNodeMsg
{
	string id   = 1; // 签名节点id Signing node ID
	bytes sign	= 2; // 签名节点签名信息 Sign node signature information
	bytes pub   = 3; // 签名者公钥 The signer's public key
}

message TxMsgInfo
{
  uint32    type = 1; // 交易流转类型 Transaction flow type
  bytes     tx = 2; // 交易信息 Transaction Information
  uint64    height = 3; // 发起交易节点高度值 The height value of the initiating transaction node
}

// PC端交易信息体 PC-side transaction information body
message TxMsgReq
{
  string       version              = 1; // 版本	version					
  TxMsgInfo    txMsgInfo            = 2;
  Vrf      vrfInfo              = 3;
  Vrf      txvrfInfo              = 4;
	repeated SignNodeMsg signNodeMsg  = 5;  // 交易流转中签名节点的签名信息 The signature information of the signing node in the transaction flow
	repeated string prevBlkHashs      = 6;	// 前置区块hashs		Pre-block hashs	

}

message Vrf
{
  string data = 1;
  CSign  Vrfsign = 2;
}

message TxMsgAck
{
  string version = 1; // 版本 version
  int32 code = 2; // 错误码 Error code
  string message = 3; // 返回错误信息 Returns an error message

  bytes tx = 4;
}

message BuildBlockBroadcastMsg
{
  string      version         = 1;                         // 版本	 			 version	
  string      id              = 2;
	bytes 		  blockRaw		    = 3;				          		// 块信息 Block information
  Vrf         vrfInfo         = 4;
  int32       flag            = 5;
  int32       type            = 6;
  repeated string      castAddrs     = 7;
}

message BuildBlockBroadcastMsgAck 
{
  string version          = 1;
  string id               = 2;
  string success          = 3;
  string blockhash		    = 4;				          		// 块信息 Block information
}


message TxPendingBroadcastMsg
{
    string version          = 1; // 版本 version
    bytes txRaw             = 2; // 交易信息 Transaction Information
	  uint64 prevBlkHeight    = 3; // 前置的高度 The height of the front
}



/***************************验证交易确认接口*************************************/
/***************************Verify the transaction confirmation interface*************************************/

enum ConfirmCacheFlag
{
  ConfirmUnknownFlag = 0;
  ConfirmTxFlag = 1;
  ConfirmRpcFlag = 2;
}

message ConfirmTransactionReq 
{
  string version = 1;
  string id = 2;
  string tx_hash = 3;
  ConfirmCacheFlag flag = 4;
}

message ConfirmTransactionAck 
{
  string version = 1;
  string id = 2;
  string tx_hash = 3;
  ConfirmCacheFlag flag = 4;
  bool success = 5;
  bytes block_raw = 6;
}



/***************************验证交易确认接口*************************************/
/***************************Verify the transaction confirmation interface*************************************/



message ConfirmTransactionIsSuccessReq 
{
  string version = 1;
  string id = 2;
  bytes block_raw = 3;
}

message ConfirmTransactionIsSuccessAck 
{
  string version = 1;
  string id = 2;
  string success = 3;
  bytes block_raw = 4;
}

message FailureBroadcastMsg
{
	string  version = 1;  // 版本 version		
	bytes   tx = 		 2;	// 交易信息 Transaction Information
  string  info = 3;
}